### PR TITLE
deriveBits and deriveKey tests

### PR DESCRIPTION
**Pull request by engelke**
New tests for subtleCrypto deriveBits and deriveKey methods for all all supported algorithms (ECDH, HKDF, PBKDF2).

### CODE REVIEW

**Comment by sideshowbarker** [comment]
@jimsch this is waiting on review

File: WebCryptoAPI/derive_bits_keys/ecdh_bits.js [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    .then(function(results) {
+        publicKeys = results.publicKeys;
+        privateKeys = results.privateKeys;
+        ecdsaKeyPairs = results.ecdsaKeyPairs;
+        noDeriveBitsKeys = results.noDeriveBitsKeys;
+
+        Object.keys(sizes).forEach(function(namedCurve) {
+            // Basic success case
+            promise_test(function(test) {
+                return subtle.deriveBits({name: "ECDH", namedCurve: namedCurve, public: publicKeys[namedCurve]}, privateKeys[namedCurve], 8 * sizes[namedCurve])
```
Comment (jimsch) [inline WebCryptoAPI/derive_bits_keys/ecdh_bits.js]: namedCurve does not appear to be a used field.  Should it be removed?

File: WebCryptoAPI/derive_bits_keys/ecdh_keys.js [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    .then(function(results) {
+        publicKeys = results.publicKeys;
+        privateKeys = results.privateKeys;
+        ecdsaKeyPairs = results.ecdsaKeyPairs;
+        noDeriveKeyKeys = results.noDeriveKeyKeys;
+
+        Object.keys(sizes).forEach(function(namedCurve) {
+            // Basic success case
+            promise_test(function(test) {
+                return subtle.deriveKey({name: "ECDH", namedCurve: namedCurve, public: publicKeys[namedCurve]}, privateKeys[namedCurve], {name: "AES-CBC", length: 128}, true, ["encrypt", "decrypt"])
```
Comment (jimsch) [inline WebCryptoAPI/derive_bits_keys/ecdh_keys.js]: This does not need to be fixed, but the idea of using a block cipher here just makes me unhappy from the cryptographers point of view.  The result of an ECDH operation is not really random.  I would feel happier if you used  HMAC-SHA-256 here instead.   This should be the more common case.

File: WebCryptoAPI/derive_bits_keys/hkdf.js [hunk 3]
```diff
@@ -0,0 +1,294 @@
+
+function run_test() {
+    // May want to test prefixed implementations.
+    var subtle = self.crypto.subtle;
+
+    // hkdf2_vectors sets up test data with the correct derivations for each
+    // test case.
+    var testData = getTestData();
+    var passwords = testData.passwords;
```
Comment (jimsch) [inline WebCryptoAPI/derive_bits_keys/hkdf.js]: You don't have to fix, but i would prefer a different term here than password.  HKDF is only marginally defined to be a password based KDF function.  PBKDF is better for that.  Not misleading people is always good.

File: WebCryptoAPI/derive_bits_keys/ecdh_bits.js (hunk 1 above)
Comment (engelke) [inline WebCryptoAPI/derive_bits_keys/ecdh_bits.js]: Yes it should, and has been for deriveBits now.

Comment (engelke) [inline WebCryptoAPI/derive_bits_keys/ecdh_bits.js]: And now also for deriveKeys.

File: WebCryptoAPI/derive_bits_keys/ecdh_keys.js (hunk 2 above)
Comment (engelke) [inline WebCryptoAPI/derive_bits_keys/ecdh_keys.js]: Changed.

File: WebCryptoAPI/derive_bits_keys/hkdf.js (hunk 3 above)
Comment (engelke) [inline WebCryptoAPI/derive_bits_keys/hkdf.js]: Done.

**Pull request merged**